src/XamlX/Transform/WhitespaceNormalization.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using XamlX.Ast;
+
+namespace XamlX.Transform
+{
+    internal static class WhitespaceNormalization
+    {
+        public static bool IsWhitespace(string text)
+        {
+            foreach (var t in text)
+            {
+                if (!IsWhitespace(t))
+                {
+                    return false;
+                }
+            }
+
+            // NOTE: empty text nodes count as whitespace
+            return true;
+        }
+
+        private static bool IsWhitespace(char ch)
+        {
+            return ch == ' ' || ch == '\n' || ch == '\t';
```

Comment:
We should use a Unicode conform test here:
https://docs.microsoft.com/en-us/dotnet/api/system.char.iswhitespace?view=net-5.0#System_Char_IsWhiteSpace_System_Char_

Reply:
This is actually deliberate. Both the XAML Spec and the documentation I cited above specifically define XAML-whitespace as only these three characters. WPF also only checks for these three characters specifically. (See for example: https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/Parser/XamlText.cs#L188)

Reply:
I would not trust that dated code. Where does the normalization happen?

Reply:
The \r normalization should already happen in the XML parser as per the XML spec (section 2.11): https://www.w3.org/TR/xml/

I do have a test for this in `CarriageReturnIsTranslatedToNewLineEvenWithXmlSpacePreserve` to make sure this even applies for `xml:space="preserve"`.

There's a case to be made for including `\r` in the definition above in case someone goes crazy and includes an XML character entity for a carriage return (as in &#xD;)

Reply:
I've added `\r` to the recognized whitespace and added tests to cover because `&#x0D;` does indeed bypass the XML parser's normalization.

Reply:
What happens if a copy pasted string contains zero width spaces, non breaking spaces etc.? Are those trimmed?

Reply:
No, but that is consistent with WPF.